lib/shaders/pathtracer_compute.wgsl

```diff
@@ -85,190 +84,139 @@ fn traceRay(ray: Ray, seed: u32) -> vec3f {
 }
 
-fn hitFloor(ray: Ray, floor: Floor) -> f32 {
-    var t: f32 = -(dot(ray.origin, floor.normal) + floor.distance) / dot(ray.direction, floor.normal);
-    if (t > 0.0) {
-        return t;
-    }
-    return -1.0;
-}
+fn intersectSphere(
+    sphere: ptr<function, Sphere>,
+    material: ptr<function, Material>,
+    ray: Ray,
+    bestHit: ptr<function, HitInfo>) {
 
-fn hitSphere(ray: Ray, sphere: Sphere) -> f32 {
-    
     let a: f32 = dot(ray.direction, ray.direction);
-    let b: f32 = 2.0 * dot(ray.direction, ray.origin - sphere.center);
-    let c: f32 = dot(ray.origin - sphere.center, ray.origin - sphere.center) - sphere.radius * sphere.radius;
+    let b: f32 = 2.0 * dot(ray.direction, ray.origin - (*sphere).center);
+    let c: f32 = dot(ray.origin - (*sphere).center, ray.origin - (*sphere).center) - (*sphere).radius * (*sphere).radius;
     let discriminant: f32 = b * b - 4.0 * a * c;
 
     if(discriminant < 0.0) {
-        return -1.0;
+        return;
     }
 
     let t0: f32 = (-b - sqrt(discriminant)) / (2.0 * a);
     let t1: f32 = (-b + sqrt(discriminant)) / (2.0 * a);
 
-    return min(t0, t1);
+    let t: f32 = min(t0, t1);
+
+    if (t > 0.0 && t < (*bestHit).t) {
+        (*bestHit).t = t;
+        (*bestHit).normal = normalize(rayAt(ray, t) - (*sphere).center);
+        (*bestHit).material = (*material); // TODO: reference directly or use index
+    }
 }
 
-fn hitTriangle(ray: Ray, v0: vec3<f32>, v1: vec3<f32>, v2: vec3<f32>) -> f32 {
-    const EPSILON: f32 = 0.0000001;
-    
-    var vertex0: vec3<f32> = v0;
-    var vertex1: vec3<f32> = v1;
-    var vertex2: vec3<f32> = v2;
+fn intersectTriangles(ray: Ray, bestHit: ptr<function, HitInfo>) {
+    for(var i: u32 = 0; i < 8712; i+=9) { // TODO: pass triangle count as uniform 
 
-    var edge1: vec3<f32> = vertex1 - vertex0;
-    var edge2: vec3<f32> = vertex2 - vertex0;
-    var rayVecXe2: vec3<f32> = cross(ray.direction, edge2);
+        var v0: vec3f = vec3f(vertices[i], vertices[i+1], vertices[i+2]);
+        var v1: vec3f = vec3f(vertices[i+3], vertices[i+4], vertices[i+5]);
+        var v2: vec3f = vec3f(vertices[i+6], vertices[i+7], vertices[i+8]);
 
-    var det: f32 = dot(edge1, rayVecXe2);
+        const EPSILON: f32 = 0.0000001;
+    
+        var vertex0: vec3f = v0;
+        var vertex1: vec3f = v1;
+        var vertex2: vec3f = v2;
 
-    if (det > -EPSILON && det < EPSILON) {
-        return -1; // This ray is parallel to this triangle.
-    }
+        var edge1: vec3f = vertex1 - vertex0;
+        var edge2: vec3f = vertex2 - vertex0;
+        var rayVecXe2: vec3f = cross(ray.direction, edge2);
 
-    var invDet: f32 = 1.0 / det;
-    var s: vec3<f32> = ray.origin - vertex0;
-    var u: f32 = invDet * dot(s, rayVecXe2);
+        var det: f32 = dot(edge1, rayVecXe2);
 
-    if (u < 0.0 || u > 1.0) {
-        return -1;
-    }
+        if (det > -EPSILON && det < EPSILON) { // This ray is parallel to this triangle.
+            return; 
```

Comment:
Here algorithm returns and not checking all triangle intersections. Instead it should continue.